packages/run-protocol/src/vpool-xyk-amm/addLiquidity.js

```diff
@@ -28,3 +34,143 @@ export const makeMakeAddLiquidityInvitation = (zcf, getPool) => {
 
   return makeAddLiquidityInvitation;
 };
+
+/**
+ * The pool has poolX and poolY currently. The user wants to add liquidity of
+ * giveX and giveY, but the ratios are probably not the same. We want to adjust
+ * the pool to have the ratio (poolX + giveX) / (poolY + giveY), without
+ * changing K, the product of the two sides.
+ *
+ * Calculate targetX and targetY, which multiply to the same product as
+ * poolX * poolY, but are in the ratio we'll end at. From this we can produce
+ * a proposed trade that maintains K, but changes the ratio, so we can add
+ * liquidity at the desired ratio.
```

Comment:
I don't trust myself to check arithmetic... let alone floating point arithmetic. So I tried some property-based testing in #4533 . Those tests **generated failing cases** so either:

 - I don't understand this spec well enough to translate it into property tests, or
 - This implementation is not to spec.

a couple examples of failures:

```
  with giveX=5 giveY=990668278, targetX=0 and targetY=35689 multiply to the same product as poolX=9 * poolY=2.

  Difference:

  - 0n
  + 18n
```

```
  with giveX=5 giveY=999999996, targetX=3 and targetY=587666676 multiply to the same product as poolX=5 * poolY=469899091.

  Difference:

  - 1763000028n
  + 2349495455n
```

Reply:
After looking at the failures, I decided that the main cause was instances where the pool's original liquidity was in single digits, which meant that the pool didn't have enough resolution to allow changing the pool ratio closely enough to be helpful. The easy solution is just to forbid this cases, and return a 'no trade' indication.

I've updated this PR, and also "improved" the parameter based test to better show the intent of the change. The original param-based test diagnosed it as a failure when the K value of the pool changed or the ratio of X to Y changed. I changed the tests to allow for those values to change by a few percent. When the pre-existing pool balance and the amount being added are relatively large, the discrepancy is small. 

When the starting pool balance is in single digits, the final values don't need to exactly match.